### ISSUE TITLE
ember-cli-babel in peerDependencies for addons

### ISSUE THREAD

**Issue by jlami** [ember-cli/ember-cli#7399]
Hi,

I was wondering if it is possible to change the way addons specify their dependency on ember-cli-babel. Currently it is a hard dependency. And in my case the different versions that addons specify result in many different copies in node_modules. I believe this can be reduced by chaning it into peerDependencies. https://github.com/ember-cli/ember-addon-output/blob/c7874269a64a6a8edf2436de6265a274853cfcc7/package.json#L20
Normally peerDependencies have the tendency to confuse users, as they are not automatically installed. But ember-new-output already has a devDependency on it, so new ember-cli projects should have it.
(edit: for addon development there should also be a devDependency on it)

The only problem I see, is that addons currently specify a very narrow range. ^5.1.7 for example. This will not match the version that is currently in a new ember project (6+). So maybe the ember-addon-output should specify a `>5`?

I'm not completely aware of how the babel invocation is handled. So I'm not sure if an addon that specified `ember-cli-babel: ^5` will get version 5 in a project that has another addon which requires 6. I would expect ember-cli to actually run one version of babel over all the addons, but maybe I'm missing some concern here? But I would expect an addon to still work with a higher version of babel.

**Comment by rwjblue** [comment]
I completely agree with you. I wrote up https://github.com/ember-cli/rfcs/pull/112 to enable his sort of thing a few days ago.

**Comment by jlami** [comment]
Ah, ok. What a close timing :). Didn't think of looking in the rfcs. Should I close this issue? This is more of a specific case that could be addressed when things are working, so maybe a reminder here is good?

**Comment by rwjblue** [comment]
Yeah, do you mind reviewing and chiming in on the RFC?  Not a lot of folks with positive thoughts over there 😝 .

**Comment by jlami** [comment]
After looking at this some more I found out that ember-cli handles addons a bit differently than I thought. I had in my mind that it would only load one version of an ember addon. But it really builds a DAG, which wil contain the addons separately for every addon in the dep tree. Which I find a bit odd, or at least inconsistent with the output that will be generated from the flattened addon tree (and app tree). I've run into problems before where multiple versions of an addon were loaded in the dependency tree, but only one version would end up in the resulting application (and of course not the one I expected). But that is probably up for another rfc ;)

Maybe it should still be possible to use different versions of an addon if it only does internal (pre/post)processing and has no other output in the resulting addon/app tree. But I'm not really convinced. I don't really see the point in an addon specifying what kind of transpilation should be done for the parent project. I can see the need for some of the preprocessing, like sass into css and hbs into js. But to use different versions of JS transpilation by using different versions of babel? I feel the application should be more in control of that. It should specify what the target should be able to support. And the version of babel that is used to do those final transformations could be influenced by the addon requirements. But having multiple versions still feels wrong. Right now that is probably necessary, since every addon can set its own options for babel. Does it need to spread objects, then babel 6 is probably required. But maybe it should set a js-feature-rest-objects or something to indicate this. Then the project can decide whether its target already supports that or if it really does need babel to convert that feature into lower JS. Which I believe is the goal with babel 6 targets.

My attempt to use peer deps with ember-cli-babel did work when I included pkg['peerDependencies'] in the result of the addon-discovery.js dependencies function. But that will not actually check if the versions are correct, so it could still be a conflicting setup. And it still doesn't solve my concern that addons can specify their own babel settings and version.
